lib/sdk/tabs/tab-firefox.js

```diff
@@ -3,309 +3,307 @@
     deprecateUsage(
       'tab.favicon is deprecated, ' +
       'please use require("sdk/places/favicon").getFavicon instead.'
     );
-    return this._tab ? getFaviconURIForLocation(this.url) : undefined
+    return isClosed(this) ? undefined : getFaviconURIForLocation(this.url);
+  },
+
+  get contentType() {
+    return isClosed(this) ? undefined : browser(this).documentContentType;
+  },
+
+  get index() {
+    return isClosed(this) ? undefined : getIndex(viewsFor.get(this));
+  },
+
+  set index(val) {
+    if (isClosed(this))
+      return;
+
+    move(viewsFor.get(this), val);
   },
-  /**
-   * The CSS style for the tab
-   */
-  get style() null, // TODO
-  /**
-   * The index of the tab relative to other tabs in the application window.
-   * Changing this property will change order of the actual position of the tab.
-   * @type {Number}
-   */
-  get index()
-    this._tab ?
-    this._window.gBrowser.getBrowserIndexForDocument(this._contentDocument) :
-    undefined,
-  set index(value)
-    this._tab && this._window.gBrowser.moveTabTo(this._tab, value),
-  /**
-   * Thumbnail data URI of the page currently loaded in this tab.
-   * @type {String}
-   */
-  getThumbnail() {
-    if (!this._tab)
+
+  get isPinned() {
+    return isClosed(this) ? undefined : viewsFor.get(this).pinned;
+  },
+
+  get window() {
+    if (isClosed(this))
       return undefined;
-    if (this._tab.getAttribute('remote')) {
-      console.error('This method is not supported with E10S');
-      return BLANK;
-    }
-    return getThumbnailURIForWindow(this._contentWindow);
+
+    require('../windows');
+    let tabElement = viewsFor.get(this);
+    let domWindow = tabElement.ownerDocument.defaultView;
+    return modelFor(domWindow);
+  },
+
+  get readyState() {
+    // Not e10s compliant
+    return isClosed(this) ? undefined : browser(this).contentDocument.readyState;
   },
-  /**
-   * Whether or not tab is pinned (Is an app-tab).
-   * @type {Boolean}
-   */
-  get isPinned() this._tab ? this._tab.pinned : undefined,
-  pin: function pin() {
-    if (!this._tab)
+
+  pin: function() {
+    if (isClosed(this))
       return;
-    this._window.gBrowser.pinTab(this._tab);
+
+    pin(viewsFor.get(this));
   },
-  unpin: function unpin() {
-    if (!this._tab)
+
+  unpin: function() {
+    if (isClosed(this))
       return;
-    this._window.gBrowser.unpinTab(this._tab);
+
+    unpin(viewsFor.get(this));
   },
 
-  /**
-   * Create a worker for this tab, first argument is options given to Worker.
-   * @type {Worker}
-   */
-  attach: function attach(options) {
-    if (!this._tab)
+  close: function(callback) {
+    let tabElement = viewsFor.get(this);
+
+    if (!tabElement || !tabElement.parentNode) {
+      callback();
       return;
-    // BUG 792946 https://bugzilla.mozilla.org/show_bug.cgi?id=792946
-    // TODO: fix this circular dependency
-    let { Worker } = require('./worker');
-    return Worker(options, this._contentWindow);
```

Comment:
I guess we can close [bug 792946](https://bugzilla.mozilla.org/show_bug.cgi?id=792946) now.